src/Components/Web.JS/src/Rendering/JSRootComponents.ts

```diff
@@ -7,100 +7,170 @@ let nextPendingDynamicRootComponentIdentifier = 0;
+    const containerIdentifier = pendingRootComponentContainerNamePrefix + (++nextPendingDynamicRootComponentIdentifier).toString();
+    pendingRootComponentContainers.set(containerIdentifier, toElement);
+
+    // Instruct .NET to add and render the new root component
+    const componentId = await getRequiredManager().invokeMethodAsync<number>(
+      'AddRootComponent', componentIdentifier, containerIdentifier);
+    const component = new DynamicRootComponent(componentId, jsComponentParametersByIdentifier[componentIdentifier]);
+    await component.setParameters(initialParameters);
+    return component;
+  },
 };
 
 export function getAndRemovePendingRootComponentContainer(containerIdentifier: string): Element | undefined {
-    const container = pendingRootComponentContainers.get(containerIdentifier);
-    if (container) {
-        pendingRootComponentContainers.delete(containerIdentifier);
-        return container;
-    }
+  const container = pendingRootComponentContainers.get(containerIdentifier);
+  if (container) {
+    pendingRootComponentContainers.delete(containerIdentifier);
+    return container;
+  }
+}
+
+class EventCallbackWrapper {
+  private _wrapper: { func: any };
+
+  private readonly _jsObjectReference: any;
+
+  constructor(eventCallback: any) {
+    this._wrapper = { func: eventCallback };
+    this._jsObjectReference = DotNet.createJSObjectReference(this._wrapper);
+  }
+
+  setCallback(eventCallback: any) {
+    this._wrapper.func = eventCallback;
+  }
+
+  getJSObjectReference() {
+    return this._jsObjectReference;
+  }
+
+  dispose() {
+    DotNet.disposeJSObjectReference(this._jsObjectReference);
+  }
 }
 
 class DynamicRootComponent {
-    private _componentId: number | null;
+  private _componentId: number | null;
 
-    constructor(componentId: number) {
-        this._componentId = componentId;
-    }
+  private _jsEventCallbackWrappers = new Map<string, EventCallbackWrapper | null>();
+
+  constructor(componentId: number, parameters: JSComponentParameter[]) {
+    this._componentId = componentId;
 
-    setParameters(parameters: ComponentParameters) {
-        parameters = parameters || {};
-        const parameterCount = Object.keys(parameters).length;
-        return getRequiredManager().invokeMethodAsync('SetRootComponentParameters', this._componentId, parameterCount, parameters);
+    for (const parameter of parameters) {
+      if (parameter.type === 'eventcallback') {
+        // We lazily initialize event callback wrappers, so null entries are added for now.
+        this._jsEventCallbackWrappers.set(parameter.name.toLowerCase(), null);
+      }
+    }
+  }
+
+  setParameters(parameters: ComponentParameters) {
+    parameters = parameters || {};
+
+    const keys = Object.keys(parameters);
+    const parameterCount = keys.length;
+
+    // We iterate over the provided parameters to modify event callback values.
+    for (const key of keys) {
```

Comment:
Not sure whether this is a helpful suggestion or not, so please feel free to disregard if it makes things less convenient but:

Since the eventcallback parameters are likely to be a small subset of the total number of parameters, would it be faster and easier to make an array of eventcallback parameter names during the constructor, and then iterate over that here? Then we don't have to walk through any unrelated parameters. It might also no longer be necessary to prepopulate `_jsEventCallbackWrappers` with null values.

Reply:
Update: this suggestion might be invalidated by [this](https://github.com/dotnet/aspnetcore/pull/35206/files?w=1#r686645716), but it's an open question and I'm certainly happy for you to decide how to balance these tradeoffs!